cdap-tms/src/main/java/co/cask/cdap/messaging/client/ClientMessagingService.java

```diff
@@ -0,0 +1,511 @@
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.Encoder;
+import org.apache.avro.io.EncoderFactory;
+import org.apache.tephra.TransactionCodec;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.InetSocketAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * The client implementation of {@link MessagingService}. This client is intended for internal
+ * higher level API implementation only.
+ *
+ * NOTE: This class shouldn't expose to end user (e.g. cdap-client module).
+ */
+public final class ClientMessagingService implements MessagingService {
+
+  // A arbitrary timeout for getting endpoints from discovery service.
+  // It is mainly for the first discovery call as it takes time for ZK to mirror the changes.
+  private static final long DISCOVERY_PICK_TIMEOUT_SECS = 5L;
+  private static final HttpRequestConfig HTTP_REQUEST_CONFIG = new DefaultHttpRequestConfig();
+  private static final TransactionCodec TRANSACTION_CODEC = new TransactionCodec();
+  private static final Gson GSON = new Gson();
+  // These types for only for Gson to use, hence using the gson TypeToken instead of guava one
+  private static final Type TOPIC_PROPERTY_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final Type TOPIC_LIST_TYPE = new TypeToken<List<String>>() { }.getType();
+
+  private final EndpointStrategy endpointStrategy;
+
+  @VisibleForTesting
+  @Inject
+  public ClientMessagingService(DiscoveryServiceClient discoveryServiceClient) {
+    this.endpointStrategy = new RandomEndpointStrategy(
+      discoveryServiceClient.discover(Constants.Service.MESSAGING_SERVICE));
+  }
+
+  @Override
+  public void createTopic(TopicMetadata topicMetadata) throws TopicAlreadyExistsException, IOException {
+    TopicId topicId = topicMetadata.getTopicId();
+
+    HttpRequest request = HttpRequest
+      .put(createURL(createTopicPath(topicId)))
+      .withBody(GSON.toJson(topicMetadata.getProperties()))
+      .build();
+    HttpResponse response = HttpRequests.execute(request, HTTP_REQUEST_CONFIG);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_CONFLICT) {
+      throw new TopicAlreadyExistsException(topicId);
```

Comment:
It can't. This is because the handleError cannot throw `TopicAlreadyExistsException` or `TopicNotFoundException` due to the fact that it's been called by both `createTopic` and `updateTopic` methods.